File: src/optimizers/idpp.jl
# ==============================================================================
# IDPP and S-IDPP Path Interpolation
# ==============================================================================
#
# Image Dependent Pair Potential (IDPP) and Sequential IDPP (S-IDPP) methods
# for generating initial NEB paths. These produce much better initial guesses
# than linear interpolation by ensuring pairwise distances vary smoothly.
#
# Reference:
#   Smidstrup, S., Pedersen, A., Stokbro, K. & Jonsson, H. (2014).
#   Improved initial guess for minimum energy path calculations.
#   J. Chem. Phys., 140, 214106.
#
#   Schmerwitz, Y. L. A., Gunde, M., Goswami, R. & Jonsson, H. (2024).
#   Improved initialization of the optimized path of NEB transition state
#   searches. arXiv:2407.16810v2.

# ==============================================================================
# Per-image IDPP
# ==============================================================================

"""
    idpp_interpolation(x_start, x_end, n_images; kwargs...) -> Vector{Vector{Float64}}

Create an IDPP-interpolated path of `n_images` images (including fixed
endpoints). Each intermediate image is independently optimized so that its
pairwise distances match a linearly interpolated target.

Matches eOn's `idppPath` in `NEBInitialPaths.cpp`.
"""
function idpp_interpolation(
    x_start::Vector{Float64},
    x_end::Vector{Float64},
    n_images::Int;
    n_coords_per_atom::Int = 3,
    max_iter::Int = 5000,
    max_move::Float64 = 0.1,
    force_tol::Float64 = 0.001,
    lbfgs_memory::Int = 20,
)
    images = linear_interpolation(x_start, x_end, n_images)
    n_atoms = div(length(x_start), n_coords_per_atom)

    d_start = _pairwise_distances(x_start, n_atoms, n_coords_per_atom)
    d_end = _pairwise_distances(x_end, n_atoms, n_coords_per_atom)

    for img_idx in 2:(n_images - 1)
        xi = (img_idx - 1) / (n_images - 1)
        d_target = (1 - xi) * d_start + xi * d_end

        optim = OptimState(lbfgs_memory)
        x = copy(images[img_idx])

        for iter in 1:max_iter
            _, force = _idpp_energy_force(x, d_target, n_atoms, n_coords_per_atom)
            _max_atom_force(force, n_atoms, n_coords_per_atom) < force_tol && break
            disp = optim_step!(optim, x, force, max_move;
                               n_coords_per_atom)
            x .+= disp
        end

        images[img_idx] = x
    end

    return images
end

# ==============================================================================
# S-IDPP (Sequential IDPP)
# ==============================================================================

"""
    sidpp_interpolation(x_start, x_end, n_images; kwargs...) -> Vector{Vector{Float64}}

Create an S-IDPP path. Grows the path incrementally from both endpoints,
running collective IDPP-NEB relaxation after each growth step, followed by
a final full-path relaxation.

Matches eOn's `sidppPath` in `NEBInitialPaths.cpp`.
"""
function sidpp_interpolation(
    x_start::Vector{Float64},
    x_end::Vector{Float64},
    n_images::Int;
    n_coords_per_atom::Int = 3,
    max_iter::Int = 5000,
    max_move::Float64 = 0.1,
    force_tol::Float64 = 0.001,
    lbfgs_memory::Int = 20,
    spring_constant::Float64 = 5.0,
    growth_alpha::Float64 = 0.33,
)
    n_atoms = div(length(x_start), n_coords_per_atom)

    d_init = _pairwise_distances(x_start, n_atoms, n_coords_per_atom)
    d_final = _pairwise_distances(x_end, n_atoms, n_coords_per_atom)

    # Start with [reactant, product]
    path = [copy(x_start), copy(x_end)]

    n_target = n_images - 2
    n_left = 0
    n_right = 0
    n_intermediate = 0

    # Growth loop: alternate left and right
    while n_intermediate < n_target
        if n_intermediate < n_target
            frontier = path[n_left + 1]
            next = path[n_left + 2]
            insert!(path, n_left + 2,
                    (1 - growth_alpha) * frontier + growth_alpha * next)
            n_left += 1
            n_intermediate += 1
        end

        if n_intermediate < n_target
            right_idx = length(path) - n_right
            frontier = path[right_idx]
            prev = path[right_idx - 1]
            insert!(path, right_idx,
                    (1 - growth_alpha) * frontier + growth_alpha * prev)
            n_right += 1
            n_intermediate += 1
        end

        _relax_collective_idpp!(
            path, d_init, d_final, n_atoms, n_coords_per_atom;
            max_iter, max_move, force_tol, lbfgs_memory, spring_constant,
        )
    end

    # Final full-path relaxation
    _relax_collective_idpp!(
        path, d_init, d_final, n_atoms, n_coords_per_atom;
        max_iter = 500, max_move, force_tol, lbfgs_memory, spring_constant,
    )

    return path
end

# ==============================================================================
# Collective IDPP-NEB relaxation (shared by SIDPP growth steps)
# ==============================================================================

function _relax_collective_idpp!(
    path::Vector{Vector{Float64}},
    d_init::Matrix{Float64},
    d_final::Matrix{Float64},
    n_atoms::Int,
    n_coords::Int;
    max_iter::Int = 5000,
    max_move::Float64 = 0.1,
    force_tol::Float64 = 0.001,
    lbfgs_memory::Int = 20,
    spring_constant::Float64 = 5.0,
)
    n_images = length(path)
    n_mov = n_images - 2
    n_mov == 0 && return
    D = length(path[1])

    optim = OptimState(lbfgs_memory)

    for iter in 1:max_iter
        forces = _collective_idpp_forces(
            path, d_init, d_final, n_atoms, n_coords, spring_constant,
        )

        cur_force = vcat(forces[2:end-1]...)
        _max_atom_force(cur_force, div(length(cur_force), n_coords), n_coords) < force_tol && break

        cur_x = vcat(path[2:end-1]...)
        disp = optim_step!(optim, cur_x, cur_force, max_move;
                           n_coords_per_atom = n_coords)

        new_x = cur_x + disp
        for i in 1:n_mov
            off = (i - 1) * D
            path[i + 1] = new_x[off+1:off+D]
        end
    end
end

# Perpendicular IDPP forces + spring forces parallel to tangent.
function _collective_idpp_forces(
    path::Vector{Vector{Float64}},
    d_init::Matrix{Float64},
    d_final::Matrix{Float64},
    n_atoms::Int,
    n_coords::Int,
    k_spring::Float64,
)
    n_images = length(path)
    D = length(path[1])
    forces = [zeros(D) for _ in 1:n_images]

    for i in 2:(n_images - 1)
        xi = (i - 1) / (n_images - 1)
        d_target = (1 - xi) * d_init + xi * d_final

        _, f_idpp = _idpp_energy_force(path[i], d_target, n_atoms, n_coords)

        # Simple tangent: (next - prev), normalized
        tau = path[i + 1] - path[i - 1]
        tn = norm(tau)
        tau = tn > 1e-18 ? tau / tn : zeros(D)

        # Perpendicular IDPP + parallel spring
        f_perp = f_idpp - dot(f_idpp, tau) * tau
        d_next = norm(path[i + 1] - path[i])
        d_prev = norm(path[i] - path[i - 1])
        f_spring = k_spring * (d_next - d_prev) * tau

        forces[i] = f_perp + f_spring
    end

    return forces
end

# ==============================================================================
# IDPP primitives
# ==============================================================================

function _pairwise_distances(x::Vector{Float64}, n_atoms::Int, n_coords::Int)
    d = zeros(n_atoms, n_atoms)
    for i in 1:n_atoms
        off_i = (i - 1) * n_coords
        for j in (i + 1):n_atoms
            off_j = (j - 1) * n_coords
            r = norm(@view(x[off_i+1:off_i+n_coords]) - @view(x[off_j+1:off_j+n_coords]))
            d[i, j] = r
            d[j, i] = r
        end
    end
    return d
end

# E = 0.5 * sum_{i<j} (1/r^4) * (r - d_target)^2
function _idpp_energy_force(
    x::Vector{Float64},
    d_target::Matrix{Float64},
    n_atoms::Int,
    n_coords::Int,
)
    energy = 0.0
    force = zeros(length(x))

    for i in 1:n_atoms
        off_i = (i - 1) * n_coords
        for j in (i + 1):n_atoms
            off_j = (j - 1) * n_coords

            dr = x[off_i+1:off_i+n_coords] - x[off_j+1:off_j+n_coords]
            r = norm(dr)
            r = max(r, 1e-4)

            diff = r - d_target[i, j]
            r4 = r^4

            energy += 0.5 * diff^2 / r4

            # dE/dr = diff * (1 - 2*diff/r) / r^4
            dEdr = diff * (1.0 - 2.0 * diff / r) / r4
            f_pair = (-dEdr / r) .* dr

            force[off_i+1:off_i+n_coords] .+= f_pair
            force[off_j+1:off_j+n_coords] .-= f_pair
        end
    end

    return energy, force
end

# Max per-atom force norm
function _max_atom_force(force::Vector{Float64}, n_atoms::Int, n_coords::Int)
    max_f = 0.0
    for a in 1:n_atoms
        off = (a - 1) * n_coords
        max_f = max(max_f, norm(@view force[off+1:off+n_coords]))
    end
    return max_f
end

File: crates/chemgp-core/src/idpp.rs
//! IDPP and S-IDPP path interpolation.
//!
//! Ports `idpp.jl`: Image Dependent Pair Potential for generating initial
//! NEB paths with smooth pairwise distance variation.
//!
//! Reference: Smidstrup et al., J. Chem. Phys. 140, 214106 (2014).

use crate::neb_path::linear_interpolation;
use crate::optim_step::OptimState;

/// Configuration for IDPP/S-IDPP path interpolation.
#[derive(Clone, Copy)]
pub struct IdppConfig {
    pub n_images: usize,
    pub n_coords_per_atom: usize,
    pub max_iter: usize,
    pub max_move: f64,
    pub force_tol: f64,
    pub lbfgs_memory: usize,
}

/// IDPP interpolation: per-image independent optimization.
pub fn idpp_interpolation(
    x_start: &[f64],
    x_end: &[f64],
    cfg: &IdppConfig,
) -> Vec<Vec<f64>> {
    let IdppConfig { n_images, n_coords_per_atom, max_iter, max_move, force_tol, lbfgs_memory } = *cfg;
    let mut images = linear_interpolation(x_start, x_end, n_images);
    let n_atoms = x_start.len() / n_coords_per_atom;

    let d_start = pairwise_distances(x_start, n_atoms, n_coords_per_atom);
    let d_end = pairwise_distances(x_end, n_atoms, n_coords_per_atom);

    for (img_idx, image) in images.iter_mut().enumerate().take(n_images - 1).skip(1) {
        let xi = img_idx as f64 / (n_images - 1) as f64;
        let d_target: Vec<f64> = d_start
            .iter()
            .zip(d_end.iter())
            .map(|(&a, &b)| (1.0 - xi) * a + xi * b)
            .collect();

        let mut optim = OptimState::new(lbfgs_memory);
        let mut x = image.clone();

        for _ in 0..max_iter {
            let (_, force) =
                idpp_energy_force(&x, &d_target, n_atoms, n_coords_per_atom);
            if max_atom_force(&force, n_atoms, n_coords_per_atom) < force_tol {
                break;
            }
            let disp = optim.step(&x, &force, max_move, n_coords_per_atom);
            for j in 0..x.len() {
                x[j] += disp[j];
            }
        }

        *image = x;
    }

    images
}

/// S-IDPP: sequential growth from both endpoints with collective relaxation.
pub fn sidpp_interpolation(
    x_start: &[f64],
    x_end: &[f64],
    cfg: &IdppConfig,
    spring_constant: f64,
    growth_alpha: f64,
) -> Vec<Vec<f64>> {
    let n_images = cfg.n_images;
    let n_coords_per_atom = cfg.n_coords_per_atom;
    let n_atoms = x_start.len() / n_coords_per_atom;

    let d_init = pairwise_distances(x_start, n_atoms, n_coords_per_atom);
    let d_final = pairwise_distances(x_end, n_atoms, n_coords_per_atom);

    let mut path = vec![x_start.to_vec(), x_end.to_vec()];
    let n_target = n_images.saturating_sub(2);
    let mut n_left = 0;
    let mut n_right = 0;
    let mut n_intermediate = 0;

    while n_intermediate < n_target {
        if n_intermediate < n_target {
            let frontier = &path[n_left];
            let next = &path[n_left + 1];
            let new_img: Vec<f64> = frontier
                .iter()
                .zip(next.iter())
                .map(|(&a, &b)| (1.0 - growth_alpha) * a + growth_alpha * b)
                .collect();
            path.insert(n_left + 1, new_img);
            n_left += 1;
            n_intermediate += 1;
        }

        if n_intermediate < n_target {
            let right_idx = path.len() - 1 - n_right;
            let frontier = path[right_idx].clone();
            let prev = path[right_idx - 1].clone();
            let new_img: Vec<f64> = frontier
                .iter()
                .zip(prev.iter())
                .map(|(&a, &b)| (1.0 - growth_alpha) * a + growth_alpha * b)
                .collect();
            path.insert(right_idx, new_img);
            n_right += 1;
            n_intermediate += 1;
        }

        relax_collective_idpp(
            &mut path, &d_init, &d_final, n_atoms, cfg, spring_constant,
        );
    }

    // Final full-path relaxation
    let final_cfg = IdppConfig { max_iter: 500, ..*cfg };
    relax_collective_idpp(
        &mut path, &d_init, &d_final, n_atoms, &final_cfg, spring_constant,
    );

    path
}

/// Collective IDPP-NEB relaxation.
fn relax_collective_idpp(
    path: &mut [Vec<f64>],
    d_init: &[f64],
    d_final: &[f64],
    n_atoms: usize,
    cfg: &IdppConfig,
    spring_constant: f64,
) {
    let IdppConfig { n_coords_per_atom: n_coords, max_iter, max_move, force_tol, lbfgs_memory, .. } = *cfg;
    let n_images = path.len();
    let n_mov = if n_images >= 2 { n_images - 2 } else { return };
    let d = path[0].len();

    let mut optim = OptimState::new(lbfgs_memory);

    for _ in 0..max_iter {
        let forces =
            collective_idpp_forces(path, d_init, d_final, n_atoms, n_coords, spring_constant);

        let mut cur_force = Vec::with_capacity(n_mov * d);
        for f in forces.iter().take(n_mov + 1).skip(1) {
            cur_force.extend_from_slice(f);
        }

        let total_atoms = cur_force.len() / n_coords;
        if max_atom_force(&cur_force, total_atoms, n_coords) < force_tol {
            break;
        }

        let mut cur_x = Vec::with_capacity(n_mov * d);
        for p in path.iter().take(n_mov + 1).skip(1) {
            cur_x.extend_from_slice(p);
        }

        let disp = optim.step(&cur_x, &cur_force, max_move, n_coords);
        let new_x: Vec<f64> = cur_x.iter().zip(disp.iter()).map(|(a, b)| a + b).collect();

        for i in 0..n_mov {
            let off = i * d;
            path[i + 1] = new_x[off..off + d].to_vec();
        }
    }
}

/// Perpendicular IDPP forces + spring forces parallel to tangent.
fn collective_idpp_forces(
    path: &[Vec<f64>],
    d_init: &[f64],
    d_final: &[f64],
    n_atoms: usize,
    n_coords: usize,
    k_spring: f64,
) -> Vec<Vec<f64>> {
    let n_images = path.len();
    let d = path[0].len();
    let mut forces: Vec<Vec<f64>> = (0..n_images).map(|_| vec![0.0; d]).collect();

    for i in 1..n_images - 1 {
        let xi = i as f64 / (n_images - 1) as f64;
        let d_target: Vec<f64> = d_init
            .iter()
            .zip(d_final.iter())
            .map(|(&a, &b)| (1.0 - xi) * a + xi * b)
            .collect();

        let (_, f_idpp) = idpp_energy_force(&path[i], &d_target, n_atoms, n_coords);

        // Simple tangent: (next - prev), normalized
        let tau_raw: Vec<f64> = path[i + 1]
            .iter()
            .zip(path[i - 1].iter())
            .map(|(a, b)| a - b)
            .collect();
        let tn: f64 = tau_raw.iter().map(|x| x * x).sum::<f64>().sqrt();
        let tau: Vec<f64> = if tn > 1e-18 {
            tau_raw.iter().map(|x| x / tn).collect()
        } else {
            vec![0.0; d]
        };

        // Perpendicular IDPP + parallel spring
        let f_dot_t: f64 = f_idpp.iter().zip(tau.iter()).map(|(f, t)| f * t).sum();
        let d_next: f64 = path[i + 1]
            .iter()
            .zip(path[i].iter())
            .map(|(a, b)| (a - b).powi(2))
            .sum::<f64>()
            .sqrt();
        let d_prev: f64 = path[i]
            .iter()
            .zip(path[i - 1].iter())
            .map(|(a, b)| (a - b).powi(2))
            .sum::<f64>()
            .sqrt();

        for j in 0..d {
            let f_perp = f_idpp[j] - f_dot_t * tau[j];
            let f_spr = k_spring * (d_next - d_prev) * tau[j];
            forces[i][j] = f_perp + f_spr;
        }
    }

    forces
}

/// Pairwise distances as a flat upper-triangular vector.
pub fn pairwise_distances(x: &[f64], n_atoms: usize, n_coords: usize) -> Vec<f64> {
    let mut d = vec![0.0; n_atoms * n_atoms];
    for i in 0..n_atoms {
        for j in i + 1..n_atoms {
            let r: f64 = (0..n_coords)
                .map(|k| (x[i * n_coords + k] - x[j * n_coords + k]).powi(2))
                .sum::<f64>()
                .sqrt();
            d[i * n_atoms + j] = r;
            d[j * n_atoms + i] = r;
        }
    }
    d
}

/// IDPP energy and force: E = 0.5 * sum_{i<j} (1/r^4) * (r - d_target)^2.
fn idpp_energy_force(
    x: &[f64],
    d_target: &[f64],
    n_atoms: usize,
    n_coords: usize,
) -> (f64, Vec<f64>) {
    let mut energy = 0.0;
    let mut force = vec![0.0; x.len()];

    for i in 0..n_atoms {
        for j in i + 1..n_atoms {
            let dr: Vec<f64> = (0..n_coords)
                .map(|k| x[i * n_coords + k] - x[j * n_coords + k])
                .collect();
            let r: f64 = dr.iter().map(|d| d * d).sum::<f64>().sqrt().max(1e-4);

            let diff = r - d_target[i * n_atoms + j];
            let r4 = r.powi(4);

            energy += 0.5 * diff * diff / r4;

            let de_dr = diff * (1.0 - 2.0 * diff / r) / r4;
            for k in 0..n_coords {
                let f = -de_dr / r * dr[k];
                force[i * n_coords + k] += f;
                force[j * n_coords + k] -= f;
            }
        }
    }

    (energy, force)
}

/// Max per-atom force norm.
fn max_atom_force(force: &[f64], n_atoms: usize, n_coords: usize) -> f64 {
    let mut max_f = 0.0f64;
    for a in 0..n_atoms {
        let off = a * n_coords;
        let f: f64 = (0..n_coords).map(|d| force[off + d].powi(2)).sum::<f64>().sqrt();
        max_f = max_f.max(f);
    }
    max_f
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_idpp_preserves_endpoints() {
        let start = vec![0.0, 0.0, 0.0, 1.0, 0.0, 0.0];
        let end = vec![0.0, 0.0, 0.0, 2.0, 0.0, 0.0];
        let images = idpp_interpolation(&start, &end, &IdppConfig {
            n_images: 5, n_coords_per_atom: 3, max_iter: 100, max_move: 0.1, force_tol: 0.01, lbfgs_memory: 10,
        });
        assert_eq!(images.len(), 5);
        assert!((images[0][3] - 1.0).abs() < 1e-10);
        assert!((images[4][3] - 2.0).abs() < 1e-10);
    }
}
